### PR TITLE
Write integration tests for citizen api

### PR DESCRIPTION
**Pull request by Shrutika-Ingale**
Writing integration tests for citizen APIs

### CODE REVIEW

File: src/Kmd.Momentum.Mea.Api/appsettings.json [hunk 1]
```diff
@@ -8,8 +8,8 @@
     "SignUpSignInPolicyId": "B2C_1A_signup_signin"
   },
   "MeaClient": {
-    "BaseAddress": "https://kmd-rct-momentum-159-api.azurewebsites.net/api/",
-    "Timeout": "00:00:30"
+    "BaseAddress": "https://kmd-rct-momentum-159-api.azurewebsites.net/api",
+    "Timeout": "00:10:30"
```
Comment (ExploreMyCode) [inline src/Kmd.Momentum.Mea.Api/appsettings.json]: Any reason to update above setting?

File: src/Kmd.Momentum.Mea.Common/MeaHttpClient/MeaClient.cs [hunk 2]
```diff
@@ -22,11 +23,12 @@ public async Task<string> GetAsync(Uri url)
             var authResponse = await ReturnAuthorizationTokenAsync().ConfigureAwait(false);
 
             var accessToken = JObject.Parse(await authResponse.Content.ReadAsStringAsync().ConfigureAwait(false))["access_token"];
-            _httpClient.DefaultRequestHeaders.Add("Authorization", $"Bearer {(string)accessToken}");
+            _httpClient.DefaultRequestHeaders.Clear();
+            _httpClient.DefaultRequestHeaders.Authorization = AuthenticationHeaderValue.Parse("bearer " + accessToken);
```
Comment (ExploreMyCode) [inline src/Kmd.Momentum.Mea.Common/MeaHttpClient/MeaClient.cs]: I think, Amit has change this file. You need to take latest.

File: src/Kmd.Momentum.Mea/Citizen/CitizenService.cs [hunk 3]
```diff
@@ -1,8 +1,11 @@
...
 using Kmd.Momentum.Mea.MeaHttpClientHelper;
 using Microsoft.Extensions.Configuration;
+using Newtonsoft.Json;
 using Newtonsoft.Json.Linq;
 using System;
+using System.Collections.Generic;
+using System.Linq;
 using System.Threading.Tasks;
 
 namespace Kmd.Momentum.Mea.Citizen
```
Comment (ExploreMyCode) [inline src/Kmd.Momentum.Mea/Citizen/CitizenService.cs]: Why this file is coming as part of your PR? This file has concrete implementation of service. You shouldn't change it. Any reason ?

File: src/Kmd.Momentum.Mea/Citizen/ICitizenService.cs [hunk 4]
```diff
@@ -1,11 +1,12 @@
 ﻿using Kmd.Momentum.Mea.Citizen.Model;
+using System.Collections.Generic;
```
Comment (ExploreMyCode) [inline src/Kmd.Momentum.Mea/Citizen/ICitizenService.cs]: Same here. This file shouldn't come as part of your changes.

File: src/Kmd.Momentum.Mea.Api/appsettings.json (hunk 1 above)
Comment (Shrutika-Ingale) [reply src/Kmd.Momentum.Mea.Api/appsettings.json]: It has come from dev branch. I haven't modified it.

File: test/Kmd.Momentum.Mea.Tests/Citizen/CitizenTests.cs [hunk 5]
```diff
@@ -15,35 +18,36 @@ public class CitizenTests
...
-#pragma warning restore CA2000 // Dispose objects before losing scope
+             var _configuration = new Mock<IConfiguration>();
+            _configuration.SetupGet(x => x["KMD_MOMENTUM_MEA_McaApiUri"]).Returns("http://google.com/");
+
+            var mockResponseData = new List<string>();
 
-            var citizenService = new CitizenService(helperHttpClientMoq.Object, _configurationRoot.Object);
+            mockResponseData.Add(JsonConvert.SerializeObject(new CitizenDataResponseModel("testId1", "TestDisplay1","givenname","middlename","initials","test@email.com","1234567891","","description",true,true)));
+            mockResponseData.Add(JsonConvert.SerializeObject(new CitizenDataResponseModel("testId2", "TestDisplay2", "givenname", "middlename", "initials", "test@email.com", "1234567891", "", "description", true, true)));
+            helperHttpClientMoq.Setup(x => x.GetAllActiveCitizenDataFromMomentumCoreAsync(new Uri($"{_configuration.Object["KMD_MOMENTUM_MEA_McaApiUri"]}/search"))).Returns(Task.FromResult((IReadOnlyList<string>)mockResponseData));
```
Comment (ExploreMyCode) [inline test/Kmd.Momentum.Mea.Tests/Citizen/CitizenTests.cs]: You need to add mockResponseData into collection.

File: src/Kmd.Momentum.Mea.Common/MeaHttpClient/MeaClient.cs (hunk 2 above)
Comment (Shrutika-Ingale) [reply src/Kmd.Momentum.Mea.Common/MeaHttpClient/MeaClient.cs]: The branch is merged with latest code now.

File: src/Kmd.Momentum.Mea/Citizen/CitizenService.cs (hunk 3 above)
Comment (Shrutika-Ingale) [reply src/Kmd.Momentum.Mea/Citizen/CitizenService.cs]: The branch is merged with latest code now.

File: src/Kmd.Momentum.Mea/Citizen/ICitizenService.cs (hunk 4 above)
Comment (Shrutika-Ingale) [reply src/Kmd.Momentum.Mea/Citizen/ICitizenService.cs]: The branch is merged with latest code now.

File: src/Kmd.Momentum.Mea/Citizen/Model/CitizenDataResponseModel.cs [hunk 6]
```diff
@@ -1,19 +1,45 @@
-﻿namespace Kmd.Momentum.Mea.Citizen.Model
```
[the remark this answers is not in the document]
Comment (Shrutika-Ingale) [reply src/Kmd.Momentum.Mea/Citizen/Model/CitizenDataResponseModel.cs]: The branch is merged with latest code now.

File: src/Kmd.Momentum.Mea/MeaHttpClientHelper/HttpClientHelper.cs [hunk 7]
```diff
@@ -12,19 +15,75 @@ public class HttpClientHelper : IHttpClientHelper
         public HttpClientHelper(IMeaClient meaClient)
         {
             _meaClient = meaClient;
-        }        
+        }
 
-        public async Task<string[]> GetAllActiveCitizenDataFromMomentumCoreAsync(Uri url)
+        public async Task<IReadOnlyList<string>> GetAllActiveCitizenDataFromMomentumCoreAsync(Uri url)
         {
-            var response = await _meaClient.GetAsync(url).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (Shrutika-Ingale) [reply src/Kmd.Momentum.Mea/MeaHttpClientHelper/HttpClientHelper.cs]: The branch is merged with latest code now.

File: test/Kmd.Momentum.Mea.Integration.Tests/Citizens/CitizenTests.cs [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+                services.AddScoped(_ => meaHttpClientMoq.Object);
+            }));
+
+            var clientMoq = mockedFactory.CreateClient();
+
+            var mockResponseData = new List<string>();
+
+            List<CitizenDataResponseModel> lst = new List<CitizenDataResponseModel>()
+            {
+                new CitizenDataResponseModel("testId1", "TestDisplay1", "givenname", "middlename", "initials",
```
[the remark this answers is not in the document]
Comment (Shrutika-Ingale) [reply test/Kmd.Momentum.Mea.Integration.Tests/Citizens/CitizenTests.cs]: Okay..

File: test/Kmd.Momentum.Mea.Integration.Tests/Citizens/CitizenTests.cs [hunk 9]
```diff
@@ -0,0 +1,136 @@
+﻿using FluentAssertions;
+using Kmd.Momentum.Mea.Api;
```
[the remark this answers is not in the document]
Comment (Shrutika-Ingale) [reply test/Kmd.Momentum.Mea.Integration.Tests/Citizens/CitizenTests.cs]: yes

File: test/Kmd.Momentum.Mea.Integration.Tests/Citizens/CitizenTests.cs [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+        [Fact]
+        public async Task GetCitizenByCprNoSuccess()
+        {
+            //Arrange
+            var cprNumber = "0208682105";
+            var requestUri = $"/citizens/cpr/{cprNumber}";
+            var httpClientHelperMoq = new Mock<IHttpClientHelper>();
+
+            var citizenDataResponse = new CitizenDataResponseModel("testId1", "TestDisplay1", "givenname", "middlename", "initials", "test@email.com", "1234567891", "", "description");
```
[the remark this answers is not in the document]
Comment (Shrutika-Ingale) [reply test/Kmd.Momentum.Mea.Integration.Tests/Citizens/CitizenTests.cs]: okay

File: test/Kmd.Momentum.Mea.Tests/Citizen/CitizenTests.cs (hunk 5 above)
Comment (Shrutika-Ingale) [reply test/Kmd.Momentum.Mea.Tests/Citizen/CitizenTests.cs]: The branch is merged with latest code now.

File: azure-pipelines-release-template.yml [hunk 11]
```diff
@@ -1,6 +1,7 @@
 parameters:
   azureSubscription: 'KMD Momentum Internal (77a2aca7-3b91-463d-9165-c6d0bb3689f9)'
   environmentNameSuffix: phoenix
+  environment: Phonix
```
Comment (Ashish0410verma) [inline azure-pipelines-release-template.yml]: Phoenix?

File: azure-pipelines.yml [hunk 12]
```diff
@@ -14,13 +14,14 @@ stages:
...
       vmImage: $(imageName)
+    
     steps:
     - task: UseDotNet@2
       inputs:
         packageType: 'sdk'
         useGlobalJson: true
     - task: NuGetAuthenticate@0
-    - pwsh: ./build.ps1 -PublishArtifactsToAzureDevOps:(("$(imageName)" -eq "windows-2019")) -ArtifactsStagingPath "$(Build.ArtifactStagingDirectory)"
+    - pwsh: ./build.ps1 -PublishArtifactsToAzureDevOps:(("$(imageName)" -eq "windows-2019")) -ArtifactsStagingPath "$(Build.ArtifactStagingDirectory)" -ClientId "$(McaClientId)" -ClientSecret "$(McaClientSecret)" -McaAppUri "$(McaApiUri)" -Scope "$(McaScope)" -Environment "$(ENVIRONMENT)"
```
Comment (Ashish0410verma) [inline azure-pipelines.yml]: Here it should be -McaAppUri "$(McaApiUri) => McaApiUri ?

File: build.ps1 [hunk 13]
```diff
@@ -137,8 +157,12 @@ try {
         Push-Location "$testFolder"
         try {
             Write-Host "build: Testing project in '$testFolder'"
-
-            & dotnet test -c Release --logger trx --verbosity="$BuildVerbosity" --no-build --no-restore
+            
+            ($env:KMD_MOMENTUM_MEA_ClientSecret = $ClientSecret); 
+            ($env:KMD_MOMENTUM_MEA_ClientId = $ClientId); 
+            ($env:KMD_MOMENTUM_MEA_McaApiUri = $McaAppUri);
```
Comment (Ashish0410verma) [inline build.ps1]: $McaAppUri -> $McaApiUri ? Please confirm once.

File: build.ps1 [hunk 14]
```diff
@@ -58,7 +58,27 @@ Param(
...
+    [Parameter(Mandatory=$true)]
+    $ClientId,
+
+    # The client secret for integration tests to run
+    [Parameter(Mandatory=$true)]
+    $ClientSecret,
+
+    # The MCA Api Uri for integration tests to run
+    [Parameter(Mandatory=$true)]
+    $McaAppUri,
```
Comment (Ashish0410verma) [inline build.ps1]: $McaApiUri?

**Comment by Ashish0410verma** [comment]
I think mistakenly binary and cache files are checked-in. Please remove them.

File: test/Kmd.Momentum.Mea.Integration.Tests/IntegrationConfig.cs [hunk 15]
```diff
@@ -0,0 +1,23 @@
+﻿using Microsoft.Extensions.Configuration;
+using System;
```
Comment (Ashish0410verma) [inline test/Kmd.Momentum.Mea.Integration.Tests/IntegrationConfig.cs]: Please remove unnecessary using blocks. Make sure the unused code statements are not checked in.

File: azure-pipelines-release-template.yml (hunk 11 above)
Comment (Shrutika-Ingale) [reply azure-pipelines-release-template.yml]: okay

File: test/Kmd.Momentum.Mea.Integration.Tests/IntegrationConfig.cs [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Kmd.Momentum.Mea.Integration.Tests
+{
+    public class IntegrationConfig
+    {
+        public static string GetEnvironmentName() =>
```
Comment (Ashish0410verma) [inline test/Kmd.Momentum.Mea.Integration.Tests/IntegrationConfig.cs]: Is there any reason of creating this class. Can't we do the same in IntegrationTestApplicationFactory class?

Comment (Ashish0410verma) [reply test/Kmd.Momentum.Mea.Integration.Tests/IntegrationConfig.cs]: Unused using blocks?

File: test/Kmd.Momentum.Mea.Integration.Tests/SkipLocalFactAttribute.cs [hunk 17]
```diff
@@ -0,0 +1,30 @@
+﻿using Microsoft.Extensions.Configuration;
+using System;
+using System.Collections.Generic;
+using System.Linq;
```
Comment (Ashish0410verma) [inline test/Kmd.Momentum.Mea.Integration.Tests/SkipLocalFactAttribute.cs]: unused using blocks?

File: test/Kmd.Momentum.Mea.Integration.Tests/SkipLocalFactAttribute.cs [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
+    public class SkipLocalFactAttribute: FactAttribute
+    {
+        private static IConfigurationRoot deployedConfigJson = IntegrationConfig.CreateConfigBuilder().Build();
+
+        public SkipLocalFactAttribute()
+        {
+            const string localApiUriKey = "MeaUri";
+            const string localhost = "localhost";
+            var logicApiUrl = deployedConfigJson.GetValue<Uri>(localApiUriKey, defaultValue: null);
```
Comment (Ashish0410verma) [inline test/Kmd.Momentum.Mea.Integration.Tests/SkipLocalFactAttribute.cs]: variable names have logic prefix. Please change them to mea

File: test/Kmd.Momentum.Mea.Integration.Tests/IntegrationConfig.cs (hunk 15 above)
Comment (Shrutika-Ingale) [reply test/Kmd.Momentum.Mea.Integration.Tests/IntegrationConfig.cs]: ok

File: azure-pipelines.yml (hunk 12 above)
Comment (amitxagarwal) [reply azure-pipelines.yml]: i didn't get this comment

Comment (Ashish0410verma) [reply azure-pipelines.yml]: McaAppUri or McaApiUri that's my question here?

**Pull request merged**